src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -0,0 +1,740 @@
+	}
+}
+
+type metricsAppender struct {
+	metricsAppenderOptions
+
+	tags                 *tags
+	multiSamplesAppender *multiSamplesAppender
+}
+
+type metricsAppenderOptions struct {
+	agg                     aggregator.Aggregator
+	clockOpts               clock.Options
+	tagEncoder              serialize.TagEncoder
+	matcher                 matcher.Matcher
+	encodedTagsIteratorPool *encodedTagsIteratorPool
+}
+
+func (a *metricsAppender) AddTag(name, value string) {
+	a.tags.names = append(a.tags.names, name)
+	a.tags.values = append(a.tags.values, name)
+}
+
+func (a *metricsAppender) SamplesAppender() (SamplesAppender, error) {
+	// Sort tags
+	sort.Sort(a.tags)
+
+	// Encode tags and compute a temporary (unowned) ID
+	a.tagEncoder.Reset()
+	if err := a.tagEncoder.Encode(a.tags); err != nil {
+		return nil, err
+	}
+	data, ok := a.tagEncoder.Data()
+	if !ok {
+		return nil, fmt.Errorf("unable to encode tags: names=%v, values=%v",
+			a.tags.names, a.tags.values)
+	}
+
+	unownedID := data.Bytes()
+
+	// Match policies and rollups and build samples appender
+	id := a.encodedTagsIteratorPool.Get()
+	id.Reset(unownedID)
+	now := time.Now()
+	fromNanos, toNanos := now.Add(-1*a.clockOpts.MaxNegativeSkew()).UnixNano(),
+		now.Add(1*a.clockOpts.MaxPositiveSkew()).UnixNano()
+	matchResult := a.matcher.ForwardMatch(id, fromNanos, toNanos)
+	id.Close()
+
+	policies := matchResult.MappingsAt(now.UnixNano())
+
+	a.multiSamplesAppender.reset()
+	a.multiSamplesAppender.addSamplesAppender(samplesAppender{
+		agg:       a.agg,
+		unownedID: unownedID,
+		policies:  policies,
+	})
+
+	numRollups := matchResult.NumRollups()
+	for i := 0; i < numRollups; i++ {
+		rollup, ok := matchResult.RollupsAt(i, now.UnixNano())
+		if !ok {
+			continue
+		}
+
+		a.multiSamplesAppender.addSamplesAppender(samplesAppender{
+			agg:       a.agg,
+			unownedID: rollup.ID,
+			policies:  rollup.PoliciesList,
+		})
+	}
+
+	return a.multiSamplesAppender, nil
+}
+
+func (a *metricsAppender) Reset() {
+	a.tags.names = a.tags.names[:0]
+	a.tags.values = a.tags.values[:0]
+}
+
+func (a *metricsAppender) Finalize() {
+	a.tagEncoder.Finalize()
+	a.tagEncoder = nil
+}
+
+type samplesAppender struct {
+	agg       aggregator.Aggregator
+	unownedID []byte
+	policies  policy.PoliciesList
+}
+
+func (a samplesAppender) AppendCounterSample(value int64) error {
+	sample := unaggregated.MetricUnion{
+		Type:       unaggregated.CounterType,
+		OwnsID:     false,
+		ID:         a.unownedID,
+		CounterVal: value,
+	}
+	return a.agg.AddMetricWithPoliciesList(sample, a.policies)
+}
+
+func (a samplesAppender) AppendGaugeSample(value float64) error {
+	sample := unaggregated.MetricUnion{
+		Type:     unaggregated.GaugeType,
+		OwnsID:   false,
+		ID:       a.unownedID,
+		GaugeVal: value,
+	}
+	return a.agg.AddMetricWithPoliciesList(sample, a.policies)
+}
+
+// Ensure multiSamplesAppender implements SamplesAppender
+var _ SamplesAppender = (*multiSamplesAppender)(nil)
```

Comment:
Similar but one allocates and the other doesn't, hah.  Although it will probably be optimized away by the compiler, so maybe in practice it would be the same.